crates/solver/src/solver/naive_solver/multi_order_solver.rs

```diff
@@ -148,18 +148,36 @@ fn solve_with_uniswap(
     // Because the smart contracts round in the favour of the traders, it could
     // be that we actually require a bit more from the Uniswap pool in order to
     // pay out all proceeds. We move the rounding error to the sell token so that
-    // it either comes out of the fees or existing buffers. Compute that amount:
-    let uniswap_out_with_rounding = big_int_to_u256(&settlement.trade_executions().fold(
+    // it either comes out of the fees or existing buffers. It is also important
+    // to use the **synthetic** order amounts for this, as this output amount
+    // needs to be computed for what is actually intented to be traded against
+    // an AMM and not how the order will be executed (including things like
+    // surplus fees).
+    let uniswap_out_with_rounding = big_int_to_u256(&orders.iter().try_fold(
```

Comment:
I don't really understand the transformation that took place here. I have some intuition why this makes sense and it's plausible, but I'm not very confident how the PR description and change of comment above, lead to the change of code here.

Could you clarify that more?

Reply:
I'll try my best 🙈.

Basically, given `[Orders]` and `Pool`, we compute the *exact* input and output amount that satisfies the original OBA equation (basically, clearing prices are the executed Uniswap pool price).

However, the equations were originally created assuming real executed amount. Since executed amounts can't be fractional **and** the smart contract always rounds up in favour of the trader, we have:
```
(
  ∑ executed_buy_amount_of_uniswap_out_token -
  ∑ executed_sell_amount_of_uniswap_out_token
) ∈ [uniswap_out_amount, uniswap_out_amount + n_orders)
```

Additionally, when we encode the Uniswap interaction, we use `swapTokensForExactTokens`, meaning that we only ever swap exactly `uniswap_out_amount` and slippage is paid for in the input token. This means that, in some cases, we may be missing a few wei of `uniswap_out_token` which can cause settlements to revert in case of no buffers in that token.

To address this, we compute the sum above and set the actual output amount to be that, so we are never short a wei or two.

What is critical here, is that the OBA equation and trade executions are applied to the **synthetic** orders and not the real orders. This is what the change is about.